src/main.cpp

```diff
@@ -6975,33 +6987,20 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
 
     else if (strCommand == "verack")
     {
-        pfrom->SetRecvVersion(min(pfrom->nVersion, PROTOCOL_VERSION));
+        LOCK(cs_main);
+        CNodeState* state = State(pfrom->GetId());
+        if (state->fCurrentlyConnected) {
+            // Reject duplicate verack
+            pfrom->PushMessage("reject", strCommand, REJECT_DUPLICATE, string("Duplicate verack message"));
+            Misbehaving(pfrom->GetId(), 1);
+            return false;
+        }
 
+        pfrom->SetRecvVersion(min(pfrom->nVersion, PROTOCOL_VERSION));
         // Mark this node as currently connected, so we update its timestamp later.
         if (pfrom->fNetworkNode) {
-            LOCK(cs_main);
-            State(pfrom->GetId())->fCurrentlyConnected = true;
-        }
-    }
-
-
-    // Disconnect existing peer connection when:
-    // 1. The version message has been received
-    // 2. Peer version is below the minimum version for the current epoch
-    else if (pfrom->nVersion < chainparams.GetConsensus().vUpgrades[
-        CurrentEpoch(GetHeight(), chainparams.GetConsensus())].nProtocolVersion &&
-        !(
-            chainparams.NetworkIDString() == "regtest" &&
-            !GetBoolArg("-nurejectoldversions", DEFAULT_NU_REJECT_OLD_VERSIONS)
-        )
-    ) {
-        LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
-        pfrom->PushMessage("reject", strCommand, REJECT_OBSOLETE,
-                            strprintf("Version must be %d or greater",
-                            chainparams.GetConsensus().vUpgrades[
-                                CurrentEpoch(GetHeight(), chainparams.GetConsensus())].nProtocolVersion));
-        pfrom->fDisconnect = true;
-        return false;
```

Comment:
It is incorrect to remove this code, because the current epoch can change, and in that case existing connected nodes that fail to meet the new required minimum version should be disconnected.